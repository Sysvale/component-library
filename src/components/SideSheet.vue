<template>
	 <div
	 	id="overlay"
		v-if="value"
	 	id="overlay"
		tabindex="0"
		@click="shouldCloseOnBackdrop"
	>
		<div
			id="container"
			:class="floatClass"
		>
			<slot name="close-icon">
				<div
					class="text-right pb-2"
				>
					<x-icon
						id="close-icon"
						@click="$emit('input', !value)"
					/>
				</div>
			</slot>
			<slot/>
		</div>
	</div>
</template>

<script>
import { XIcon } from 'vue-feather-icons';
export default {
	components: {
		XIcon,
	},
	props: {
		value: {
			type: Boolean,
			default: false,
			description: `
				Prop to control if the SideSheet should be shown or not (usually by v-model)
			`,
		},
		right: {
			type: Boolean,
			default: false,
			description: `
				Prop to define that the element will be shown at right screen side.
			`,
		},
		left: {
			type: Boolean,
			default: false,
			description: `
				Prop to define that the element will be shown at left screen side (right ifs default).
			`,
		},
		noCloseOnBackdrop: {
			type: Boolean,
			default: false,
			description: `
				Prop to define if the element won't be dismissed when backdrop is click.
			`,
		},
		noCloseOnEsc: {
			type: Boolean,
			default: false,
			description: `
				Prop to define if the element won't be dismissed when esc pressed.
			`,
		}
	},

	computed: {
		floatClass() {
			if (this.left) {
				return 'left';
			}
			if (this.right) {
				return 'right';
			}
			return 'right';
		}
	},

	mounted() {
		let self = this; 

		window.addEventListener('keyup', function(ev) {
				if (ev.keyCode === 27) { // esc
					self.shouldCloseOnEsc();
				}
		});
	},

	methods: {
		shouldCloseOnBackdrop() {
			if(!this.noCloseOnBackdrop) {
				this.$emit('input', !this.value);
			}
		},
		shouldCloseOnEsc() {
			if(!this.noCloseOnEsc) {
				this.$emit('input', !this.value);
			}
		}
	},
};
</script>
<style>
	#overlay {
		position: fixed;
		width: 100%;
		height: 100%;
		top: 0;
		left: 0;
		right: 0;
		bottom: 0;
		background-color: rgba(0, 0, 0, 0.4);
		z-index: 9999;
		cursor: pointer;
	}

	#container {
		background: #fff;
		width: 30%;
		height: 100%;
		margin-top: 80px;
		padding: 20px;
		border-radius: 5px;
	}

	#close-icon {
		color: rgb(106, 117, 128, 0.75);
	}

	#close-icon:hover {
		color: rgb(106, 117, 128, 1);
	}

	.right {
		float: right;
	}

	.left {
		float: left;
	}
</style>
