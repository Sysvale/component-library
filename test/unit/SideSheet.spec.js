import { mount, createLocalVue } from '@vue/test-utils';
import BootstrapVue, { BButton } from 'bootstrap-vue';
import SideSheet from '../../src/components/SideSheet.vue';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

test('Component is mounted properly', () => {
	const wrapper = mount(SideSheet, {
		localVue,
		propsData: {
			value: true,
			right: true,
		},
	});
	expect(wrapper).toMatchSnapshot();
});

describe("Items styles test", () => {
	test('if content is left aligned', () => {
		const wrapper = mount(SideSheet, {
			localVue,
			propsData: {
				value: true,
				left: true,
			},
		});
		expect(wrapper.findAll('.left').length).toBe(1);
	});

	test('if content is right aligned', () => {
		const wrapper = mount(SideSheet, {
			localVue,
			propsData: {
				value: true,
				right: true,
			},
		});
		expect(wrapper.findAll('.right').length).toBe(1);
	});
});

describe("Behavior tests", () => {

	test('if noCloseOnBackdrop as true works', async (done) => {
		const wrapper = mount(SideSheet, {
			localVue,
			propsData: {
				value: true,
				right: true,
				noCloseOnBackdrop: true,
			},
		});

		const overlay = wrapper.find('#overlay');
		overlay.trigger('click');
		await wrapper.vm.$nextTick();
		expect(wrapper.emitted().input).toBeFalsy();
		done();
	});

	test('if noCloseOnEsc as true works', async () => {
		const wrapper = mount(SideSheet, {
			localVue,
			propsData: {
				value: true,
				right: true,
				noCloseOnEsc: true,
			},
		});

		wrapper.trigger('keyup', {
			keyCode: 27
		});

		await wrapper.vm.$nextTick();
		expect(wrapper.emitted().input).toBeFalsy();
	});
});
