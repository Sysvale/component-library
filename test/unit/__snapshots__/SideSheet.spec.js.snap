// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `<span id="side-sheet"><transition-stub name="slide-fade-right"><div tabindex="0" class="overlay"><div class="container px-0 right"><div class="text-right p-3"><svg xmlns="http://www.w3.org/2000/svg" width="24px" height="24px" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" id="close-icon" class="feather feather-x"><line x1="18" y1="6" x2="6" y2="18"></line><line x1="6" y1="6" x2="18" y2="18"></line></svg></div> </div></div></transition-stub></span>`;
